File: Day02/aoc_day_2.py
class CubeSet:
    def __init__(self, red=0, green=0, blue=0):
        self.red = red
        self.green = green
        self.blue = blue


class Game:
    def __init__(self, id, sets):
        self.id = id
        self.sets = sets


def parse_input(input_str):
    games = []
    for line in input_str.strip().split('\n'):
        game_id, sets_str = line.split(': ')
        game_id = int(game_id.replace("Game ", ""))
        sets = []
        for set_str in sets_str.split('; '):
            cubes = CubeSet()
            for cube in set_str.split(', '):
                count, color = cube.split(' ')
                count = int(count)
                if color == 'red':
                    cubes.red = count
                elif color == 'green':
                    cubes.green = count
                elif color == 'blue':
                    cubes.blue = count
            sets.append(cubes)
        games.append(Game(game_id, sets))
    return games


def is_game_possible(game, bag):
    for set in game.sets:
        if set.red > bag['red'] or set.green > bag['green'] or set.blue > bag['blue']:
            return False
    return True


def find_minimum_cubes(game):
    min_red = max(set.red for set in game.sets)
    min_green = max(set.green for set in game.sets)
    min_blue = max(set.blue for set in game.sets)
    return min_red, min_green, min_blue

def calculate_power(red, green, blue):
    return red * green * blue
        


def main():
    with open("./day_02_input.txt") as f:
        data = f.read().strip()
    
    games = parse_input(data)
    
    sum_ids = 0
    total_power = 0

    bag = {
        "red": 12,
        "green": 13,
        "blue": 14
    }

    for game in games:
        min_red, min_green, min_blue = find_minimum_cubes(game)
        power = calculate_power(min_red, min_green, min_blue)
        total_power += power
        print(f"Game {game.id}: Minimum cubes - Red: {min_red}, Green: {min_green}, Blue: {min_blue}, Power: {power}")
        if is_game_possible(game, bag):
            sum_ids += game.id

    print("Sum of possible game IDs:", sum_ids)
    print("Total power of minimum sets:", total_power)


if __name__ == "__main__":
    main()

File: Day02/aoc_day_2.rs
use std::fs;

struct CubeSet {
    red: usize,
    green: usize,
    blue: usize,
}

struct Game {
    id: usize,
    sets: Vec<CubeSet>,
}

impl CubeSet {
    fn new() -> CubeSet {
        CubeSet { red: 0, green: 0, blue: 0 }
    }
}

impl Game {
    fn parse_input(input: &str) -> Vec<Game> {
        input.lines().map(|line| {
            let parts: Vec<&str> = line.split(": ").collect();
            let id = parts[0].replace("Game ", "").parse::<usize>().unwrap();
            let sets = parts[1].split("; ")
                .map(|set_str| {
                    let mut cubes = CubeSet::new();
                    for cube in set_str.split(", ") {
                        let parts: Vec<&str> = cube.split_whitespace().collect();
                        let count = parts[0].parse::<usize>().unwrap();
                        match parts[1] {
                            "red" => cubes.red = count,
                            "green" => cubes.green = count,
                            "blue" => cubes.blue = count,
                            _ => {}
                        }
                    }
                    cubes
                })
                .collect();
            Game { id, sets }
        }).collect()
    }

    fn is_game_possible(&self, bag: &std::collections::HashMap<&str, usize>) -> bool {
        self.sets.iter().all(|set| {
            set.red <= *bag.get("red").unwrap() &&
            set.green <= *bag.get("green").unwrap() &&
            set.blue <= *bag.get("blue").unwrap()
        })
    }

    fn find_minimum_cubes(&self) -> (usize, usize, usize) {
        let min_red = self.sets.iter().map(|set| set.red).max().unwrap_or(0);
        let min_green = self.sets.iter().map(|set| set.green).max().unwrap_or(0);
        let min_blue = self.sets.iter().map(|set| set.blue).max().unwrap_or(0);
        (min_red, min_green, min_blue)
    }
}

fn calculate_power(red: usize, green: usize, blue: usize) -> usize {
    red * green * blue
}

fn main() {
    let data = fs::read_to_string("./day_02_input.txt").expect("Unable to read file");
    
    let games = Game::parse_input(&data);
    
    let mut sum_ids = 0;
    let mut total_power = 0;

    let bag = std::collections::HashMap::from([
        ("red", 12),
        ("green", 13),
        ("blue", 14),
    ]);

    for game in games {
        let (min_red, min_green, min_blue) = game.find_minimum_cubes();
        let power = calculate_power(min_red, min_green, min_blue);
        total_power += power;
        println!("Game {}: Minimum cubes - Red: {}, Green: {}, Blue: {}, Power: {}", game.id, min_red, min_green, min_blue, power);
        if game.is_game_possible(&bag) {
            sum_ids += game.id;
        }
    }

    println!("Sum of possible game IDs: {}", sum_ids);
    println!("Total power of minimum sets: {}", total_power);
}
